clean-modular-code/python-pep-8.md

```diff
@@ -14,237 +14,279 @@ kernelspec:
 
-There are many different tools that can help you write code that is PEP 8 compliant. A tool that checks the format of your code is called a linter.
+:::
 
-Some linters will reformat your code for you to match the standards. These include tools like Black. Or the autopep8 tool for Jupyter Notebook.
+:::{admonition} Terminology Review
+:class: tip
 
-Other linters will simply check your code and tell you if things need to be fixed. A few **Python** packages that perform linting are listed below.
+First, let's review some terminology associated with naming conventions.
 
-* <a href="https://pep8.readthedocs.io/en/release-1.7.x/" target="_blank">pep8</a>, a `Python` package that can help you check your code for adherence to the PEP 8 style guide.
-* <a href="https://github.com/hhatto/autopep8" target="_blank">autopep8</a>, another `Python` package that can be used to modify files to the PEP 8 style guide.
+* **Lowercase letter:** `b`
+* **Uppercase letter:** `B`
+* **lowercase:** `this is all lowercase words`
+* **snake case:** when words are separated by underscores: `lower_case_with_underscores`
+* **Uppercase:** All words are all uppercase letters: `UPPERCASE`
+* **Snake case** upper case: `UPPER_CASE_WITH_UNDERSCORES`
+* **CamelCase:** Every word is capitalized so they visually stand out: `CapitalizedWords`. This is sometimes also referred to as CapWords or StudlyCaps.
+  * Note: When using acronyms in CamelCase, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.
+* **mixedCase:** (differs from CapitalizedWords by initial lowercase character!)
+* **Capitalized_Words_With_Underscores:** This approach is not recommended. Use one convention and stick with it.
 
-`Python` community members expect that your code will adhere to the PEP 8 standard, and if it does not, they generally will not be shy to tell you that your code is not "Pythonic"!
+:::
 
 +++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-## Additional Resources
+## Best practices for importing libraries
 
-* <a href="https://www.python.org/dev/peps/pep-0008/" target="_blank">The PEP 8 Style Guide</a>
+### Import Python libraries at the top of your code
+
+It’s good practice to import all required libraries at the top of your **Python** script or in the first cell of a **Jupyter Notebook**. This helps anyone understand what packages are needed to run your code. It also follows PEP 8 conventions.
+
+### Organize your imports into groups
+
+PEP 8 recommends organizing imports in the following order:
+
+1. **Standard Library Imports**: These built-in modules come with Python, such as `os` and `glob`. You can find the full list [here](https://docs.python.org/3/library/index.html).
+2. **Third-Party Imports**: Libraries that you install via `pip`, like `numpy` and `pandas`.
+3. **Local Imports**: Code or modules specific to your project.
+
+Here’s an example following PEP 8 conventions:
+
+```python
+import os
```

Comment:
Let's alphabetize these imports.

Reply:
isort makes it a little hard to know what the rules are exactly but if we want to add the sorting rules in here, here are the most common rules (in addition to grouping and alphabetization):

- modules:
  - module-only imports (`import x`) come before imports with objects (`from x import y`)
  - within that, import statements sorted alphabetically by module name
- objects within a module import (the `y` in `from x import y`)
  - first sort by whether the first letter is capitalized or not
  - then alphabetically